examples/chat/chat.go

```diff
@@ -0,0 +1,222 @@
+	"github.com/libp2p/go-libp2p-net"
+	"github.com/libp2p/go-libp2p-peer"
+	"github.com/libp2p/go-libp2p-peerstore"
+	"github.com/libp2p/go-libp2p-swarm"
+	"github.com/libp2p/go-libp2p/p2p/host/basic"
+	"github.com/multiformats/go-multiaddr"
+)
+
+/*
+* addAddrToPeerstore parses a peer multiaddress and adds
+* it to the given host's peerstore, so it knows how to
+* contact it. It returns the peer ID of the remote peer.
+* @credit examples/http-proxy/proxy.go
+ */
+func addAddrToPeerstore(h host.Host, addr string) peer.ID {
+	// The following code extracts target's the peer ID from the
+	// given multiaddress
+	ipfsaddr, err := multiaddr.NewMultiaddr(addr)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	pid, err := ipfsaddr.ValueForProtocol(multiaddr.P_IPFS)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	peerid, err := peer.IDB58Decode(pid)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// Decapsulate the /ipfs/<peerID> part from the target
+	// /ip4/<a.b.c.d>/ipfs/<peer> becomes /ip4/<a.b.c.d>
+	targetPeerAddr, _ := multiaddr.NewMultiaddr(
+		fmt.Sprintf("/ipfs/%s", peer.IDB58Encode(peerid)))
+	targetAddr := ipfsaddr.Decapsulate(targetPeerAddr)
+
+	// We have a peer ID and a targetAddr so we add
+	// it to the peerstore so LibP2P knows how to contact it
+	h.Peerstore().AddAddr(peerid, targetAddr, peerstore.PermanentAddrTTL)
+	return peerid
+}
+
+func handleStream(s net.Stream) {
+	log.Println("Got a new stream!")
+
+	// Create a buffer stream for non blocking read and write.
+	rw := bufio.NewReadWriter(bufio.NewReader(s), bufio.NewWriter(s))
+
+	go readData(rw)
+	go writeData(rw)
+
+	// Hang forever
+	select {}
+}
+func readData(rw *bufio.ReadWriter) {
+	for {
+		str, _ := rw.ReadString('\n')
+
+		if str == "" {
+			return
+		}
+		if str != "\n" {
+			// Green console colour: 	\x1b[32m
+			// Reset console colour: 	\x1b[0m
+			fmt.Printf("\x1b[32m%s\x1b[0m> ", str)
+		}
+
+	}
+}
+
+func writeData(rw *bufio.ReadWriter) {
+	stdReader := bufio.NewReader(os.Stdin)
+
+	for {
+		fmt.Print("> ")
+		sendData, err := stdReader.ReadString('\n')
+
+		if err != nil {
+			panic(err)
+		}
+
+		rw.WriteString(fmt.Sprintf("%s\n", sendData))
+		rw.Flush()
+	}
+
+}
+
+func main() {
+
+	sourcePort := flag.Int("sp", 0, "Source port number")
+	dest := flag.String("d", "", "Dest MultiAddr String")
+	help := flag.Bool("help", false, "Display Help")
+
+	flag.Parse()
+
+	if *help {
+		fmt.Printf("This program demonstrates a simple p2p chat application using libp2p\n\n")
+		fmt.Printf("Usage: Run './chat -sp <SOURCE_PORT>' where <SOURCE_PORT> can be any port number. Now run './chat -d <MULTIADDR>' where <MULTIADDR> is multiaddress of previous listener host.\n")
+
+		os.Exit(0)
+	}
+
+	// Constant random source. This will always generate the same host ID on multiple execution.
+	r := rand.New(rand.NewSource(int64(*sourcePort)))
+
+	// Creates a new RSA key pair for this host
+	prvKey, pubKey, err := crypto.GenerateKeyPairWithReader(crypto.RSA, 1024, r)
```

Comment:
Even though this is *just* an example, I'd rather use a 2048 bit key in case someone copies this code. 1024bit RSA has been deprecated for a while.